src/model/entity/backpack.go

```diff
@@ -1,11 +1,239 @@
 package entity
 
+import (
+	"sort"
+)
+
+const (
+	BackpackDefaultCapacity uint = 9
+)
+
+type BackpackIsFullError struct{}
+
+func (BackpackIsFullError) Error() string {
+	return "Backpack is full, drop something"
+}
+
+type ItemIsNotInBackpackError struct{}
+
+func (ItemIsNotInBackpackError) Error() string {
+	return "Item is not found in the Backpack"
+}
+
 type Backpack struct {
 	Capacity  uint
 	ItemsNum  uint
-	Foods     []Food
-	Elixirs   []Elixir
-	Scrolls   []Scroll
-	Weapon    []Weapon
-	Treasures []Treasure
+	Elixirs   map[string][]Elixir
+	Scrolls   map[string][]Scroll
+	Foods     map[string][]Food
+	Weapons   map[string][]Weapon
+	Treasures uint
+}
+
+type Element struct {
+	Name string
+	Num  int
+	Item ItemLike
+}
+
+type ElementsList []Element
+
+func (b *Backpack) AddTreasure(t *Treasure) {
+	b.Treasures += t.Value
+}
+
+func (b *Backpack) AddItem(i ItemLike) error {
+	if b.ItemsNum < b.Capacity {
+		if e := IsElixir(i); e != nil {
+			b.Elixirs[e.Item.Name] = append(b.Elixirs[e.Item.Name], *e)
+		}
+
+		if s := IsScroll(i); s != nil {
+			b.Scrolls[s.Item.Name] = append(b.Scrolls[s.Item.Name], *s)
+		}
+
+		if f := IsFood(i); f != nil {
+			b.Foods[f.Item.Name] = append(b.Foods[f.Item.Name], *f)
+		}
+
+		if w := IsWeapon(i); w != nil {
+			b.Weapons[w.Item.Name] = append(b.Weapons[w.Item.Name], *w)
+		}
+
+		i.Taken()
+		b.ItemsNum++
+	} else {
+		return BackpackIsFullError{}
+	}
```

Comment:
Чтобы не делать подобные длинные `if...else` можно инвертировать условие и сделать отправку ошибки не на `else`, а на `if` и дальше `else` не будет нужен. Будет меньше вложенность и лучше читаемость кода.




```suggestion
	if b.ItemsNum >= b.Capacity {
		return BackpackIsFullError{}
	}
	
	if e := IsElixir(i); e != nil {
	  b.Elixirs[e.Item.Name] = append(b.Elixirs[e.Item.Name], *e)
	}
	...
```